### PR TITLE
Add support for postgres databases

### PR DESCRIPTION
**Pull request by markusahlstrand**
I couldn't find a way to add support for postgres without adding a breaking change due to the prototype patching in knex.. So the knexDataApiClient now exposes a posgres and a mysql version of the client.
Also separated out the index-file to a few different files to make it easier to grasp for me, not sure if it's objectively better or not.

### CODE REVIEW

**Comment by markusahlstrand** [comment]
The data-api-client has some issues with postgres support as well, so it might make sense to fix that first..

**Comment by markusahlstrand** [comment]
Had another look and the issue with posgres support seems to have been using sqlstring.format. Replaced it with a regex and at least in my simple use-case it works fine.

**Comment by markusahlstrand** [comment]
Added a few fixes to get knex migrations working as well.

File: data-api.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    driverName: "rds-data",
+
+    _driver() {
+      // Setup dataApiClient
+      return dataApiClient({
+        secretArn: this.config.connection.secretArn,
+        resourceArn: this.config.connection.resourceArn,
+        database: this.config.connection.database,
+        region: this.config.connection.region
+      });
```
Comment (cinqi) [inline data-api.js]: Would it be good to have option to pass all possible [parameters](https://github.com/jeremydaly/data-api-client#configuration-options) to data-api-client? This would also make possible local development against local-data-api

Comment (markusahlstrand) [reply data-api.js]: So, more or less just pass this.config.connection as parameters? Think that would work. Should we add som check here to ensure that the required parameters are available?

Comment (markusahlstrand) [reply data-api.js]: Went ahead and changed it to pass the config.connection. Also changed to use the newly published version of the data-api-client as the region fix seems to be in now. Haven't tested it thoroughly yet..

Comment (cinqi) [reply data-api.js]: Did a quick test with local setup and this seems to work fine :)

File: index.js [hunk 2]
```diff
@@ -1,202 +1,7 @@
...
-      obj.response = obj.response.numberOfRecordsUpdated;
-    }
-
-    return obj.response;
-  }
-});
-
-module.exports = ClientRDSDataAPI;
+module.exports = {
+  postgres,
```
Comment (mmcculloch) [inline index.js:5]: Maybe something along the lines of
```
module.exports = {
  postgresDataApiClient,
  mysqlDataApiClient,
};
```
Would be better if it's now 100% necessary to have split clients

Comment (markusmindforce) [reply index.js]: Would be much nicer to have the same client, but can't figure out a way to achieve it. Maybe someone more familiar with knex can figure it out.. 
Could be that appending DataApiClient to the call would make it clearer? Another option would be to use the factory pattern? Maybe that's even better? Then we wouldn't have to expose to two different clients. We would instead have some awkward wrapping though... but might still be better?

Comment (markusahlstrand) [reply index.js:5]: Would be much nicer to have the same client, but can't figure out a way to achieve it. Maybe someone more familiar with knex can figure it out..
Could be that appending DataApiClient to the call would make it clearer? Another option would be to use the factory pattern? Maybe that's even better? Then we wouldn't have to expose to two different clients. We would instead have some awkward wrapping though... but might still be better?

**Comment by mmcculloch** [comment]
I'm also having issues passing `null` into nullable int and uuid columns.

`BadRequestException: ERROR: invalid input syntax for integer: "null"`

This is when running knex seed data.

**Comment by mmcculloch** [comment]
> I'm also having issues passing `null` into nullable int and uuid columns.
> 
> `BadRequestException: ERROR: invalid input syntax for integer: "null"`
> 
> This is when running knex seed data.

From a quite look, this is because sqlstring is converting nulls into `'null'` rather than `NULL`. On my version of Postgres 10.8 Aurora Serverless it isn't liking this.

File: sqlstring.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+const sqlstring = require("sqlstring");
+const constants = require("./constants");
+
+function format(sql, values, dialect) {
+  if (dialect === constants.dialects.postgres) {
+    return sql.replace(/(\$\d+)/gi, binding => {
+      const index = parseInt(binding.slice(1));
+      const value = values[index - 1];
+
```
Comment (mmcculloch) [inline sqlstring.js]: Actually adding 

```
      if (value === null) {
        return null;
      }
```
Solves the nulling issue

Comment (markusahlstrand) [reply sqlstring.js]: Good catch. Think the version of sqlstring might be a bit optimistic. I'll see if I can add a test for null columns.

Comment (markusahlstrand) [reply sqlstring.js]: Added a fix very similar to what you suggested. Could you check if it works for you?

Comment (mmcculloch) [reply sqlstring.js]: I'm also looking into converting the prepared statement into something `knex.raw` can use. I'm a bit worried this approach might be vulnerable to SQL Injection

Comment (markusahlstrand) [reply sqlstring.js]: I noticed that data-api-client supports named parameters. Would it be an option to rewrite the query to use named parameters instead of passing it as a raw sql string? Guess that would help with the sql injections? Guess we still would need separate postgres/mysql implementations for this though which is a pity.

Comment (mmcculloch) [reply sqlstring.js]: By importing knex

`const knexClient = require('knex')({ client: 'postgres' });`
and applying a few changes to the query
```const sql = knexClient.raw(obj.sql.replace(/`/g, '"').replace(/\$(\d+)/g, '?'), obj.bindings).toString();```

I was able to resolve the Date, null and issues with quotes within strings. This replaces sqlstring, which is MySQL only.

Comment (markusmindforce) [reply sqlstring.js]: This sounds promising.. I'll see if I can get this in today.

Comment (markusahlstrand) [reply sqlstring.js]: This sounds promising.. I'll see if I can get this in today.

Comment (markusahlstrand) [reply sqlstring.js]: Sorry for the delay.. I added this to the ma/sqlstring branch and will test it out before merging to the master of the PR-repo

Comment (mmcculloch) [reply sqlstring.js]: the replaces replace backticks with quotes " and then the other replace changes `$1, $2, $3` etc into `?`. I thought backticks were compatible with postgres but I was getting errors so maybe not. If there's a more elegant way to convert the query that would be nice. But after that it uses `knex.raw` so should avoid any possible SQL injections.

**Comment by markusahlstrand** [comment]
@alan-cooney Any chance you can have a look at this PR? Or is it to far away from the maser branch?

**Comment by markusahlstrand** [comment]
I'll close this PR and open a new from another branch. Will publish the master of our repo as a separate npm package as it's hard for us to manage dependencies straight to github.

**Comment by alan-cooney** [comment]
@markusahlstrand apologies for the delay here - I'm also happy to add you as a contributor if that's easier?

I can review the code here but I'm away until early next year.

Alan

**Comment by markusmindforce** [comment]
Would be great if we could keep it the original repo. As I'm only using it from a postgres perspective it might be good if you could validate that it still works properly with mysql before merging it?

I created a new PR from a separate branch. Hope that works for you.

Thanks!

**Comment by alan-cooney** [comment]
Sure - happy to review in early Jan (or late Dec if I get a chance) - but if you'd prefer to be contributors just let me know and then I'll add both of you.

I'll add in our broader integration tests to the source as well, and we can extend this to postgres.
